### ISSUE TITLE
include the option of simulating spectra at a set of given input redshifts

### ISSUE THREAD

**Issue by moustakas** [desihub/desisim#104]
Requested feature from @sbailey:
```
Ability to pass in a redshift array and get back a set of templates at those redshifts.  This is needed for generating simulated data to match an input mock catalog.
```

**Comment by moustakas** [comment]
I have this basically implemented, although there's a potential infinite-loop trap I'm not sure how to deal with.  It's possible -- and indeed likely -- that the user will pass a redshift which will *never* pass the color cuts for a given spectral class.  For example, *none* of the ~8000 ELG basis templates will ever pass the nominal grz color-cuts if they are told to be placed at z=0.1 (and indeed anything below z~0.5 is highly unlikely, by design).  

@sbailey, @ladosamushia How should we deal with this in the context of #110?

**Comment by sbailey** [comment]
At minimum there should be a counter that stops after N tries to avoid an infinite loop.  Options after hitting that limit:
1. raise an exception
2. print a warning and then don't include that redshift in the output list (i.e. fewer outputs than inputs)
3. print a warning, grab another random redshift from the list and keep going, even if that means that it replicates that redshift.

I think I prefer (3) as default, with an option to do (1) instead.  (2) smells like trouble.

**Comment by dkirkby** [comment]
Once the templates and cuts are fixed, you can precompute exactly which templates pass the cuts on a grid of redshifts and save the results in a boolean array `pass[ntemplates,nz]`.
This array is actually quick to compute using `speclite.filters` since you can do the magnitude calculations in parallel for all templates.  Once you have precomputed the `pass` array, you can simulate any given redshift `z ~ z[iz]` without any iteration by randomly selecting an `it` where `pass[it, iz]` is `True`.   You can also immediately detect an impossible redshift when `!np.any(pass[:,iz])` and raise an appropriate exception.

**Comment by moustakas** [comment]
Thanks for the suggestion, @dkirkby.  In detail, what you describe is what I had originally implemented in the IDL (v1.0) version of the ELG templates.  However, those templates had static emission lines on them, whereas in the current template code each continuum template gets a unique emission-line spectrum (according to a physical model).  And this is why the templates are built in a ```for``` loop, which is not terribly slow but slightly annoying.

I could do what you suggest to 'pre-select' a set of templates that satisfy the color-cuts and redshift inputs based on just the continuum (this is only an issue for the ELGs), which would also allow impossible redshifts to be captured.  Then the code could go into the loop for the full template construction.  It's *possible*, however, that the emission lines could be strong enough to push the template into an "impossible" part of color-redshift space.  I can capture this rare occurrence, however, as @sbailey suggests by stopping the infinite loop after N iterations, where N might be a few hundred attempts.

**Comment by sbailey** [comment]
close via PR #132
